src/Chainweb/Cut/Create.hs

```diff
@@ -0,0 +1,442 @@
+--    other chains.
+--
+-- 3. This is a "normal" transition.
+--
+-- TODO: it is important that the semantics of this function corresponds to the
+-- respective validation in the module "Chainweb.Cut", in particular
+-- 'isMonotonicCutExtension'. It must not happen, that a cut passes validation
+-- that can't be further extended.
+--
+getCutExtension
+    :: HasCallStack
+    => HasChainId cid
+    => Cut
+        -- ^ the cut which is to be extended
+    -> cid
+        -- ^ the header onto which the new block is created. It is expected
+        -- that this header is contained in the cut.
+    -> Maybe CutExtension
+getCutExtension c cid = do
+
+    -- In a graph transition we wait for all chains to do the transition
+    -- to the new graph before moving ahead.
+    guard (not $ isGraphTransitionCut && isGraphTransitionPost)
+
+    -- If this is a graph transition step we have to check braiding for the
+    -- old /and/ the new graph.
+    when isGraphTransitionPre (void $ newAdjHashes parentGraph)
+
+    -- Finally, compute the adjacent hashes for the target graph.
+    as <- BlockHashRecord <$> newAdjHashes targetGraph
+
+    return CutExtension
+        { _cutExtensionCut' = c
+        , _cutExtensionParent' = ParentHeader p
+        , _cutExtensionAdjacentHashes' = as
+        }
+  where
+    p = c ^?! ixg (_chainId cid)
+    v = _chainwebVersion c
+    parentHeight = _blockHeight p
+    targetHeight = parentHeight + 1
+    parentGraph = chainGraphAt_ p parentHeight
+    targetGraph = chainGraphAt_ p targetHeight
+    isGraphTransitionPre = isGraphChange c targetHeight
+    isGraphTransitionPost = isGraphChange c parentHeight
+    isGraphTransitionCut = isTransitionCut c
+
+    -- | Try to get all adjacent hashes dependencies for the given graph.
+    --
+    newAdjHashes :: ChainGraph -> Maybe (HM.HashMap ChainId BlockHash)
+    newAdjHashes g =
+        imapM (\xcid _ -> hashForChain xcid)
+        $ HS.toMap
+        $ adjacentChainIds g p
+
+    hashForChain acid
+        -- existing chain
+        | Just b <- c ^? ixg acid = tryAdj b
+        -- new chain after graph transition
+        | targetHeight == genesisHeight v acid = Just $ genesisParentBlockHash v acid
+        | otherwise = error $ T.unpack
+            $ "getAdjacentParents: invalid cut, can't find adjacent hash for chain "
+            <> sshow acid
+            <> ". Parent: " <> encodeToText (ObjectEncoded p)
+
+    tryAdj :: BlockHeader -> Maybe BlockHash
+    tryAdj b
+
+        -- When the block is behind, we can move ahead
+        | _blockHeight b == targetHeight = Just $! _blockParent b
+
+        -- if the block is ahead it's blocked
+        | _blockHeight b + 1 == parentHeight = Nothing -- chain is blocked
+
+        -- If this is not a graph transition cut we can move ahead
+        | _blockHeight b == parentHeight = Just $! _blockHash b
+
+        -- The cut is invalid
+        | _blockHeight b > targetHeight = error $ T.unpack
```

Comment:
... and these could be `Left` instead of `error`

Reply:
At this point these cases are intentionally `error` / assertions. If these trigger, it would be a bug in the code. At runtime, if a caller would trigger any of these cases, all they could do would be to throw an error themself. The detailed error message were only for me when I was debugging the chain graph transition algorithm during development.

This may change, when this code get integrated with the the braiding validation logic in the Cut module. There it may be possible that an attacker injects bad cuts and triggers these cases, in which case we want to raise more meaningful braiding validation exceptions.

So, this will be addressed in upcoming PRs.